Genericize the CMake source generator for embedding files

illwieckz opened the pull request:
Extracted from #1842:

- https://github.com/DaemonEngine/Daemon/pull/1842

Extracted and polished. The design is now stable enough to be submitted for merging.

The bug I face in #1842 is unrelated to that code.

illwieckz commented:
# Listing files to embed:

```cmake
set(GLSL_EMBED_DIR "${ENGINE_DIR}/renderer/glsl_source")
set(GLSL_EMBED_LIST
	# Common shader libraries
	common.glsl
	common_cp.glsl
	fogEquation_fp.glsl
	shaderProfiler_vp.glsl
	shaderProfiler_fp.glsl
	# …
)
```

# Generating the embedded files, etc:

```cmake
	# Generate GLSL include files.
	daemon_embed_files("EngineShaders" "GLSL" "TEXT" "client-objects")
```

First option ("EngineShaders") is the wanted NameSpace, second option ("GLSL") is the SLUG used in `GLSL_EMBED_DIR` and `GLSL_EMBED_LIST` variable and list, the third option ("TEXT") selects the file format (selecting "TEXT" will strip the `\r` characters, selecting "BINARY" will apply no transformation), and the last option ("client-objects") is the target the generated source files contribute to.

And that's all. It's even not needed anymore to add the entry points by hand like this:


```cmake
set(RENDERERLIST
	${DAEMON_EMBEDDED_DIR}/EngineShaders.cpp
	${DAEMON_EMBEDDED_DIR}/EngineShaders.h
	${ENGINE_DIR}/renderer/BufferBind.h
	# …
)
```

VReaperV commented:
> So this will be used for the Vulkan binary shaders list?

Vulkan will use functionality that was already added in #1845, which is now rebased on this pr.

illwieckz commented:
> So this will be used for the Vulkan binary shaders list?

I don't know yet if that's usable for the Vulkan binary, but with the latest iteration the embedded file should be directly accessible as a `constexpr unsigned char*`.

> I guess the separate EMBED_DIR and EMBED_LIST thing was designed for VFS embedding; if we don't need that it seems over-complicated. Just use a single list of absolute paths like everything else.

Not only, it's easier to paste the path without having to add the prefix boiler plate. I prefer to set the DIR once than copypasting the DIR for every file.

> Note that the GLSL map that we have now is coded in a rather inefficient way. The map object stores the files as `std::string`, but the original char arrays are still there, so we have 2 copies of the files at all times. Then when you read out a file a 3rd copy is created... it would be better for the map to have a StringView or something.

Ouch that's awful!

That should be better now, the map is now a map of a name and a pair of `const unsigned char* data` pointer and `size_t size`.

> It doesn't make sense to me that some auxiliary functions return `char*` because that would be unusable for binary files. In any case I doubt we need those; just the map is a good enough interface.

I removed the `ReadFile()` function because with the latest iteration those two lines were fundamentally different and we should avoid confusion:

```c+++
const char* text_ptr = EngineShaders::ReadFile(filename);

audioFile = FS::PakPath::ReadFile(filename);
```

Our other `ReadFile()` functions (like the second one) return a `std::string` with known size, while we now embed a `const char*` without bound checking (but the size is stored right next to it).

It's unsafe to access the `const unsigned char*` array without reading the size too. For the GLSL code that expects text, we don't have to read the size because I added a null terminator at the end of every files. The size stored in the pair is the size of the original file (without the null terminator), so when copying it in a `std::string` (for example in the FileSystem code) we would not copy that extra null terminator that is only there for codes like the GLSL code reading the `const unsigned char*` array directly as a text buffer.

illwieckz commented:
Here is an example of how the current implementation works in the C++ code:

```c++
void test()
{
	// Using a file directly using its symbol:
	const char* file = (const char*) EngineShaders::common_glsl;

	// Loading a file by its name:
	const char* filename = "common.glsl";
	auto found = EngineShaders::FileMap.find(filename);
	const char* foundfile = (const char*) found->second.data;

	// Iterating the files:
	for ( auto it : EngineShaders::FileMap )
	{
		Log::Debug("Embedded file %s at address %p with size %d", it.first, it.second.data, it.second.size);
	}
}
```

illwieckz commented:
The map is just a convenience for accessing the files, the map just stores the name, the size and the pointer to the data, so I assume the compiler will drop the map when the data is accessed directly.

This should fit the need for:

- What we already do for the GLSL shaders.
- Embedding a file and loading it directly, for things like builtin images like `$white`, as an alternative to write a generator if a bit complex to write.
- Exposing a builtin pak in memory to the VFS, including exposing it to the virtualized game, with the ability to list files and load them using the existing VFS mechanism.

This code was tested with my builtin pak implementation, so the only thing I haven't tested is the direct access (more than compiling the above code) is the direct access, but that should work.

VReaperV commented on CMakeLists.txt:
```diff
@@ -85,7 +85,7 @@ if (Daemon_OUT)
 endif()
 
-include(DaemonBuildInfo)
+include(DaemonSourceGenerator)
```
Just `DaemonEmbed`? That's what I did in #1845 until I rebased on this pr.

slipher commented:
> I don't know how to make it use a `Str::StringRef()` there.

A StringRef could be constructed like `{reinterpret_cast<const char*>(byte_array), sizeof(byte_array) - 1}`

illwieckz commented:
This is accepted by the compiler:

```c++
const Str::StringRef screen_fp_glsl = reinterpret_cast<const char*>( data_screen_fp_glsl );
```

This isn't:

```c++
const Str::StringRef screen_fp_glsl = { reinterpret_cast<const char*>( data_screen_fp_glsl ),
 sizeof ( data_screen_fp_glsl ) - 1 };
```

But the first one has then `screen_fp_glsl.size()` set to `0`. It's not problematic for the GLSL code since it's treated as a null-terminated string, but I would like this code to be compatible with binary files that may contain valid non-terminating null bytes, and for this I need `object.size()` to work.

Also with the first implementation (without the size), with some more elaborated embedded file code than the GLSL one I get some errors about `embeddedFileMap_t` members not having `.second()` because of being `Str::BasicStringRef` instead of having it as `.second()`. I don't know what compiler optimization madness happens to get that error, but that's not a problem I had with my custom srtuct.

illwieckz commented:
It looks like I need to add a new `Str::StringRef` constructor for that.

illwieckz commented:
Yes, I just added a new constructor on my end (but without the assert) and it worked.
I'll use yours as it is more complete.

illwieckz commented:
I modified the code to be standalone (no dependency from `src/common`), it's better this way.

illwieckz commented:
For some unknown reasons code becomes simpler with that dedicated struct than with `Str::StringRef`.

For example when testing this code with #1842, here is what I need to do with `Str::StringRef`:

```c++
for (auto& p : builtinPakMap[pak.name])
{
	if (p.first == path)
	{
		const Str::StringRef& entry = p.second;
		std::string out;
		out.resize(entry.size());
		memcpy(&out[0], entry.data(), entry.size());
		return out;
	}
}
```

While I can do this simpler code with the simple dedicated struct:

```c++
const embeddedFileMapEntry_t& entry = builtinPakMap[pak.name][path];
std::string out;
out.resize(entry.size);
memcpy(&out[0], entry.data, entry.size);
return out;
```

With `Str::StringRef` I face very weird type errors and I would need the convoluted code to avoid them.

slipher commented:
> For example when testing this code with #1842, here is what I need to do with `Str::StringRef`:
> 
> ```c++
> for (auto& p : builtinPakMap[pak.name])
> {
> 	if (p.first == path)
> 	{
> 		const Str::StringRef& entry = p.second;
> 		std::string out;
> 		out.resize(entry.size());
> 		memcpy(&out[0], entry.data(), entry.size());
> 		return out;
> 	}
> }
> ```

StringRef has automatic conversion to string so you could do
```
	if (p.first == path)
	{
		return p.second;
	}
```

Anyway it's fine with me as long as the interface doesn't used `unsigned char`. LGTM

illwieckz commented:
So, here are usage examples of the API that is implemented in the code that has been merged:

```c++
#include "DaemonEmbeddedFiles/EngineShaders.h"

void test()
{
	// Using a file directly using its symbol:
	const char* file = EngineShaders::common_glsl.data;

	// Loading a file by its name:
	const char* filename = "common.glsl";
	auto found = EngineShaders::FileMap.find(filename);
	const char* foundfile = found->second.data;

	// Iterating the files:
	for ( auto it : EngineShaders::FileMap )
	{
		Log::Debug("Embedded file %s at address %p with size %d", it.first, it.second.data, it.second.size);
	}
}
```
